### PR TITLE
feat(carousel): defining id instead of running sequence

### LINKED ISSUE #263: Carousel: use user-defined ids / indexes instead of array indexes
Same issue as #215 and the same solution as in #262

### PR DESCRIPTION
**Pull request by sendilkumarn**
fixes #263

### CODE REVIEW

File: src/carousel/carousel.ts [hunk 1]
```diff
@@ -103,14 +105,38 @@ export class NgbCarousel implements AfterContentChecked,
...
+    let selectedId: string = this._getOtherSlide(this.activeId, true);
+    this.select(selectedId);
   }
 
   private _restartTimer() {
     this.pause();
     this.cycle();
   }
+
+  private _getSlideById(slideIdx: string): any {
```
Comment (pkozlowski-opensource) [inline src/carousel/carousel.ts]: `:NgbSlide` instead of `:any`

File: src/carousel/carousel.ts [hunk 2]
```diff
@@ -49,33 +52,32 @@ export class NgbSlide {
...
+    this.activeId = activeSlide ? activeSlide.id : (this._slides.length ? this._slides.first.id : null);
   }
 
   ngOnInit() { this.cycle(); }
 
   ngOnDestroy() { clearInterval(this._slideChangeInterval); }
 
-  select(slideIdx: number) { this._activeIdx = slideIdx; }
+  select(slideIdx: string) {
+    let selectedSlide: NgbSlide = this._getSlideById(slideIdx);
```
Comment (pkozlowski-opensource) [inline src/carousel/carousel.ts]: no need for `: NgbSlide` here, it will be inferred.

File: src/carousel/carousel.ts [hunk 3]
```diff
@@ -103,14 +105,38 @@ export class NgbCarousel implements AfterContentChecked,
...
     this.pause();
     this.cycle();
   }
+
+  private _getSlideById(slideIdx: string): any {
+    let slideWithId: NgbSlide[] = this._slides.filter(slide => slide.id === slideIdx);
+    return slideWithId.length ? slideWithId[0] : null;
+  }
+
+  private _getOtherSlide(id: string, next: boolean): string {
```
Comment (pkozlowski-opensource) [inline src/carousel/carousel.ts]: This methods looks quite complex. Could you think of writing it differently so it is easier to read / reason about? Maybe split it into 2 methods (`_getNextSlide` / `_getPrevSlide`)?

File: src/carousel/carousel.ts [hunk 4]
```diff
@@ -49,33 +52,32 @@ export class NgbSlide {
...
-
   @Input() interval: number = 5000;
   @Input() wrap: boolean = true;
   @Input() keyboard: boolean = true;
+  @Input() activeId: string;
 
   ngAfterContentChecked() {
-    // auto-correct _activeIdx that might have been set incorrectly as input
-    this._activeIdx = Math.min(Math.max(0, this._activeIdx), this._slides.length - 1);
+    let activeSlide: NgbSlide = this._getSlideById(this.activeId);
```
Comment (wesleycho) [inline src/carousel/carousel.ts]: I believe the `: NgbSlide` here is unnecessary (should be inferred)

File: src/carousel/carousel.ts [hunk 5]
```diff
@@ -103,14 +105,46 @@ export class NgbCarousel implements AfterContentChecked,
...
     this.cycle();
   }
+
+  private _getSlideById(slideIdx: string): NgbSlide {
+    let slideWithId: NgbSlide[] = this._slides.filter(slide => slide.id === slideIdx);
+    return slideWithId.length ? slideWithId[0] : null;
+  }
+
+  private _getNextSlide(id: string): string {
+    let otherSlideId: string = id;
```
Comment (wesleycho) [inline src/carousel/carousel.ts]: Maybe rename to `nextSlideId`

File: src/carousel/carousel.ts [hunk 6]
```diff
@@ -103,14 +105,46 @@ export class NgbCarousel implements AfterContentChecked,
...
+        let lastSlide: boolean = (idx === (slideArr.length - 1));
+        otherSlideId =
+            lastSlide ? (this.wrap ? slideArr[0].id : slideArr[slideArr.length - 1].id) : slideArr[idx + 1].id;
+      }
+    });
+    return otherSlideId;
+  }
+
+  private _getPrevSlide(id: string): string {
+    let otherSlideId: string = id;
```
Comment (wesleycho) [inline src/carousel/carousel.ts]: Maybe rename to `prevSlideId` for readability

File: src/carousel/carousel.ts [hunk 7]
```diff
@@ -103,14 +105,46 @@ export class NgbCarousel implements AfterContentChecked,
...
+    let slideArr: any[] = this._slides.toArray();
```
Comment (wesleycho) [inline src/carousel/carousel.ts]: I think this can be typed as `: NgbSlide[]`

File: src/carousel/carousel.ts [hunk 8]
```diff
@@ -103,14 +105,46 @@ export class NgbCarousel implements AfterContentChecked,
   }
 
   private _cycleToNext() {
-    const nextIdx = this._activeIdx + 1;
-    this.select(nextIdx >= this._slides.length ? (this.wrap ? 0 : nextIdx - 1) : nextIdx);
+    let selectedId: string = this._getNextSlide(this.activeId);
```
Comment (wesleycho) [inline src/carousel/carousel.ts]: This typing can be removed too as inferred

File: src/carousel/carousel.ts [hunk 9]
```diff
@@ -103,14 +105,46 @@ export class NgbCarousel implements AfterContentChecked,
...
     this.cycle();
   }
+
+  private _getSlideById(slideIdx: string): NgbSlide {
+    let slideWithId: NgbSlide[] = this._slides.filter(slide => slide.id === slideIdx);
+    return slideWithId.length ? slideWithId[0] : null;
+  }
+
+  private _getNextSlide(id: string): string {
+    let nextSlideId: string = id;
```
Comment (wesleycho) [inline src/carousel/carousel.ts]: No need for the typing here, it's obvious from the arguments

File: src/carousel/carousel.ts [hunk 10]
```diff
@@ -103,14 +105,46 @@ export class NgbCarousel implements AfterContentChecked,
...
+        let lastSlide: boolean = (idx === (slideArr.length - 1));
+        nextSlideId =
+            lastSlide ? (this.wrap ? slideArr[0].id : slideArr[slideArr.length - 1].id) : slideArr[idx + 1].id;
+      }
+    });
+    return nextSlideId;
+  }
+
+  private _getPrevSlide(id: string): string {
+    let prevSlideId: string = id;
```
Comment (wesleycho) [inline src/carousel/carousel.ts]: Ditto about the tying

File: src/carousel/carousel.ts [hunk 11]
```diff
@@ -103,14 +105,46 @@ export class NgbCarousel implements AfterContentChecked,
...
   }
+
+  private _getSlideById(slideIdx: string): NgbSlide {
+    let slideWithId: NgbSlide[] = this._slides.filter(slide => slide.id === slideIdx);
+    return slideWithId.length ? slideWithId[0] : null;
+  }
+
+  private _getNextSlide(id: string): string {
+    let nextSlideId: string = id;
+    let slideArr: NgbSlide[] = this._slides.toArray();
```
Comment (wesleycho) [inline src/carousel/carousel.ts]: I think the `: NgbSlide[]` may not be necessary here and below - can you try removing and seeing if it works?

Comment (sendilkumarn) [inline src/carousel/carousel.ts]: isn't it similar to line no: 118. I think it is needed here.

Comment (wesleycho) [inline src/carousel/carousel.ts]: Are they needed? `_slides` is typed as `QueryList<NgbSlide>`, and if one looks at the definition of QueryList, it looks like it should work (see https://github.com/angular/angular/blob/master/modules/%40angular/core/src/linker/query_list.ts#L82 and https://github.com/angular/angular/blob/master/modules/%40angular/core/src/linker/query_list.ts#L56-L58). I could be missing something though.

**Pull request closed without merging**